core/src/main/java/io/grpc/internal/DelayedClientTransport.java

```diff
@@ -249,14 +266,40 @@ public final void shutdownNow(Status status) {
 
   public final boolean hasPendingStreams() {
     synchronized (lock) {
-      return !pendingStreams.isEmpty();
+      return !pendingStreams.isEmpty() || !uncommittedRetriableStreams.isEmpty();
     }
   }
 
   @VisibleForTesting
   final int getPendingStreamsCount() {
     synchronized (lock) {
-      return pendingStreams.size();
+      return pendingStreams.size() + uncommittedRetriableStreams.size();
+    }
+  }
+
+  final void addUncommittedRetriableStream(RetriableStream<?> retriableStream) {
```

Comment:
Somewhere you need to handle the shutdown case.

Reply:
Done

Reply:
Just adding the if is not enough, as it doesn't impact the stream's state at all. When the channel is shutdown new RPCs should fail. The current `if` just ignores the fact that the channel is shutdown. Instead you need some behavior that will trigger `masterListener.closed()`. As a starting point you could return a boolean here.

Reply:
Return the shutdown and propagate it to `prestart()`. Didn't implement `RetriableStream` inside `DelayedClientTransport` because `newSubstream()` seems not easy here.